cpp/src/arrow/compute/kernels/mask.cc

```diff
@@ -0,0 +1,230 @@
+  builder->UnsafeAppend(std::forward<Scalar>(value));
+  return Status::OK();
+}
+
+static Status UnsafeAppend(BinaryBuilder* builder, util::string_view value) {
+  RETURN_NOT_OK(builder->ReserveData(static_cast<int64_t>(value.size())));
+  builder->UnsafeAppend(value);
+  return Status::OK();
+}
+
+static Status UnsafeAppend(StringBuilder* builder, util::string_view value) {
+  RETURN_NOT_OK(builder->ReserveData(static_cast<int64_t>(value.size())));
+  builder->UnsafeAppend(value);
+  return Status::OK();
+}
+
+// TODO(bkietz) this can be optimized
+static int64_t OutputSize(const BooleanArray& mask) {
+  auto offset = mask.offset();
+  auto length = mask.length();
+  internal::BitmapReader mask_data(mask.data()->buffers[1]->data(), offset, length);
+  int64_t size = 0;
+  for (auto i = offset; i < offset + length; ++i) {
+    if (mask.IsNull(i) || mask_data.IsSet()) {
+      ++size;
+    }
+    mask_data.Next();
+  }
+  return size;
+}
+
+template <bool AllValuesValid, bool WholeMaskValid, typename ValueArray,
+          typename OutBuilder>
+Status MaskImpl(FunctionContext*, const ValueArray& values, const BooleanArray& mask,
+                OutBuilder* builder) {
+  auto offset = mask.offset();
+  auto length = mask.length();
+  internal::BitmapReader mask_data(mask.data()->buffers[1]->data(), offset, length);
+  for (int64_t i = 0; i < mask.length(); mask_data.Next(), ++i) {
+    if (!WholeMaskValid && mask.IsNull(i)) {
+      builder->UnsafeAppendNull();
+      continue;
+    }
+    if (mask_data.IsNotSet()) {
+      continue;
+    }
+    if (!AllValuesValid && values.IsNull(i)) {
+      builder->UnsafeAppendNull();
+      continue;
+    }
+    RETURN_NOT_OK(UnsafeAppend(builder, values.GetView(i)));
+  }
+  return Status::OK();
+}
+
+template <bool AllValuesValid, typename ValueArray, typename MaskArray,
+          typename OutBuilder>
+Status UnpackMaskNullCount(FunctionContext* context, const ValueArray& values,
+                           const MaskArray& mask, OutBuilder* builder) {
+  if (mask.null_count() == 0) {
+    return MaskImpl<AllValuesValid, true>(context, values, mask, builder);
+  }
+  return MaskImpl<AllValuesValid, false>(context, values, mask, builder);
+}
+
+template <typename ValueArray, typename MaskArray, typename OutBuilder>
+Status UnpackValuesNullCount(FunctionContext* context, const ValueArray& values,
+                             const MaskArray& mask, OutBuilder* builder) {
+  if (values.null_count() == 0) {
+    return UnpackMaskNullCount<true>(context, values, mask, builder);
+  }
+  return UnpackMaskNullCount<false>(context, values, mask, builder);
+}
+
+template <typename T>
+using ArrayType = typename TypeTraits<T>::ArrayType;
+
+template <typename MaskType>
+struct UnpackValues {
+  template <typename ValueType>
+  Status Visit(const ValueType&) {
+    using OutBuilder = typename TypeTraits<ValueType>::BuilderType;
+    auto&& mask = static_cast<const ArrayType<MaskType>&>(*params_.mask);
```

Comment:
Use `checked_cast` everywhere here. Also why `auto&&`?

Reply:
`auto&&` because it was faster to type than `const auto&`.